Add support for direct http template uploading.

netservers opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
This is a new feature that allows direct http uploading of CloudStack templates.
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
lib/ansible/modules/cloud/cloudstack/cs_template.py
##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.4.0
  config file = /home/john/.ansible.cfg
  configured module search path = [u'/home/john/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible-2.4.0-py2.7.egg/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.13 (default, May 10 2017, 20:04:28) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->
As an alternative to specifying *url* or *vm*, specify a *file*, which is a path to a local template.
<!--- Paste verbatim command output below, e.g. before and after your change -->
Sample playbook:
```
---
- name: Upload Templates
  hosts: host1
  tasks:
  - local_action:
      module: cs_template
      name: "{{ item.name }}"
      display_text: "{{ item.name }}"
      file: "{{ item.file }}"
      zone: London
      hypervisor: KVM
      format: QCOW2
      is_featured: no
      is_public: yes
      is_extractable: yes
      sshkey_enabled: no
      password_enabled: no
      requires_hvm: yes
      os_type: "{{ item.type }}"
    with_items:
    - { type: "Other 2.6x Linux (32-bit)", name: "OpenWRT", file: "openwrt-x86-kvm_guest-combined-ext4.qcow2" }
```

Commit:
Add support for direct http template uploading.

resmo commented on lib/ansible/modules/cloud/cloudstack/cs_template.py:
```diff
@@ -390,6 +401,8 @@
 
 # import cloudstack common
 from ansible.module_utils.cloudstack import *
+import requests
```
please make a comment here that is okay here to depend on requests since this module depends on cs which depends on requests.

netservers replied:
Please see my last push that crossed with your comment. During testing it became clear that requests alone is not sufficient. I've followed a pattern I saw in the aws_api_gateway module for not imposing the dependency on those that don't use the new functionality.

resmo commented on lib/ansible/modules/cloud/cloudstack/cs_template.py:
```diff
@@ -495,6 +507,56 @@ def create_template(self):
+        template = self.get_template()
+        if not template:
+            self.result['changed'] = True
+            args                    = self._get_args()
```
this module is marked as "legacy" and therefore the CI is not complaining about pep8 here, but it would help if we try to be pep8 compliant.

netservers replied:
I think it would look weird if I only pep8 that one line. I'll pep8 the whole block.

resmo replied:
yep, that is what I meant, basically make sure new code is pep8 compliant

Commit:
Large templates need to be streamed from disk, not loaded into RAM.
requests_toolbelt.multipart.encoder is therefore required.

The import has been done in such as way as to not impose this requirement
on those not using the direct file upload feature.

resmo reviewed (changes requested):
some minor changes, lgtm overall. Needs manual testing.

(this module is on my list for adding integration testing)

Commit:
pep8 changes.

Commit:
Change format of requirements for file.

Commit:
Change format of requirements for file.

Commit:
Change format of requirements for file.

Commit:
Remove requirements from file option.

rhtyd commented:
@netservers please fix the conflict

gundalow commented:
@netservers 
Thank you for your PR.
As part of reviewing the backlog of PRs we are looking at PRs older PRs that haven't been updated in a while

Given that:
* PR has needed a rebase since Oct 2017
* This is a feature rather than a bug fix

Therefore I'm going to close this.

If you or anyone else wants to continue with this work then please do feel free to create a fresh PR and `@mention` the previous reviewers in here.

@rhtyd If you are interested in this perhaps you could raise a fresh PR?
